Checkout description, API and configuration (#1868)
ezsystems/developer-documentation: docs/commerce/checkout/configure_checkout.md

---
description: Configure the default checkouts process and build custom ones.
edition: commerce
---

# Configure checkout

When you work with your Commerce implementation, you can review and modify 
the checkout configuration.

!!! note "Permissions" 

>     When you modufy the workflow configuration, make sure you properly set user [permissions](../permission_use_cases.md#commerce) to the checkout component.

Reviewer: ```suggestion
    When you modify the workflow configuration, make sure you properly set user [permissions](../permission_use_cases.md#commerce) to the checkout component.
```

transition represents a separate checkout step. 

By default, the checkout process is configured to render each stage based on a separate 
set of libraries and templates.
Each checkout step is handled by a controller that you configure in workflow metadata.

Custom workflow implementations are defined under the `framework.workflows` key, 
and they must support the `Ibexa\Contracts\Checkout\Value\CheckoutInterface`. 
The default fallback workflow is `ibexa_checkout`, which is prepended at bundle level.
The `checkout.workflow` parameter is repository-aware.

To customize your configuration, place it in a YAML file, under the `framework.workflows.<your_workflow_name>` key, and reference it with `ibexa.repositories.<your_repository>.checkout.workflow: your_workflow_name`. 
The system can then identify which of your configured workflows handles the checkout process.

!!! note 

    When you modify or create a controller, to ensure that no user-data is lost, extend the `Ibexa\Bundle\Checkout\Controller\AbstractStep` controller and call the `advance()` method.

Each step configuration includes the following settings:

> - `controller` - A mandatory setting pointing to a library that governs the bahavior of the process. It contains all the required business logic and submits the whole step, so that a transition can happen.

Reviewer: ```suggestion
- `controller` - A mandatory setting pointing to a library that governs the behavior of the process. It contains all the required business logic and submits the whole step, so that a transition can happen.
```

                customer_content_type: <your_ct_identifier_for_customer> #customer by default, it is used in registration and uses given shipping/billing addresses to pre-populate address forms in "Select Address" checkout step
                shipping_methods:
                    courier:
                        name: "Courier"
                        translation_domain: "checkout" #optional, all the shipping/payment methods use this one by default
                    parcel_machine:
                        name: "Parcel machine"
                    self_pickup:
                        name: "Self pickup"
                    express_delivery:
                        name: "Express delivery"
                payment_methods:
                    credit_card:
                        name: "Credit card"
                    paypal:
                        name: "Paypal"
                    money_transfer:
                        name: "Money transfer"
```

## Shipping and billing address field configurations 

The billing and shipping addresses used in checkout are related to their respective 
values that exist in corporate account information. 

> You can configure these addresses by modifying the following code under `../ibexa/checkout/blob/main/src/bundle/DependencyInjection/Configuration/CheckoutParser.php`

Reviewer: That is actually wrong - no code sitting in the vendor can or should be adjusted. For changing the default format for addresses one should create a format within Fieldtype Address configuration e.g. `foo` and reference it under `billing_address_format` / `shipping_address_format` checkout configuration keys, as stated above.

Reviewer 2: Please review the updated description

---
description: Configure the default checkouts process and build custom ones.
edition: commerce
---

# Configure checkout

When you work with your Commerce implementation, you can review and modify 
the checkout configuration.

!!! note "Permissions" 

    When you modify the workflow configuration, make sure you properly set user [permissions](permission_use_cases.md#commerce) to the checkout component.

## Checkout workflow

> Checkout workflow is configured like other [workflows](workflow.md), and each 

Reviewer: If this refers to Ibexa Workflow component, this sentence is wrong. In checkout, we rely only on **Symfony** Workflow. No configuration from our Workflow Component can be used - it works only for content objects wheras checkout-specific objects are not content per se.

Reviewer 2: I have removed a reference to our workflow

                        translation_domain: checkout
                summary:
                    from:
                        - shipping_selected
                        - summarized
                    to: summarized
                    metadata:
                        next_step: complete
                        controller: Ibexa\Bundle\Checkout\Controller\CheckoutStep\SummaryStepController::renderStepView
                        label: 'Payment & summary'
                        translation_domain: checkout
                complete:
                    from: summarized
                    to: completed
                    metadata:
                        controller: Ibexa\Bundle\Checkout\Controller\CheckoutStep\CompleteStepController::renderCompleteView
                        label: 'Order confirmation'
                        translation_domain: checkout
```

## Configure shipping and payment methods

You can define the shipping and payment methods.
Under `ibexa.repositories.<repository_name>.checkout`, create entries that resemble 
the following example:

``` yaml 
ibexa:
    repositories:
        <repository_name>:
            checkout:

>                 workflow: <your_workflow_identifier> #optional, "ibexa_checkout" is used by default

Reviewer: This specific key should probably be mentioned in the section above, it doesn't refer to shipping/payment methods.

Reviewer 2: Definitely, fixed

- `billing` is part of the `ibexa/corporate-accounts` repository 
- `shipping` is part of the `ibexa/checkout` bundle's default configuration 

To modify address formats you can either override the existing formats or create 
custom ones.

### Override existing Address Field Type formats 

To override the existing `billing` and `shipping` address Field Type formats, under `ibexa.repositories.<repository_name>.checkout`, define the address format, 
that has been modified, for example, by adding a tax identifier to the `billing` address:

``` yaml 
ibexa_field_type_address:
    formats:
        shipping:
            country:
                default:
                    - region
                    - locality
                    - street
                    - postal_code
                    - email
                    - phone_number
                    
        billing:
            country:
                default:
                    - region
                    - locality
                    - street
                    - postal_code
                    - email
                    - phone_number
                    - tax_number

> ```

Reviewer: This sections needs to be removed. The `billing_address_format` / `shipping_address_format` configuration keys are meant to be used instead of overriding the existing formats as they may be used in other parts of the system, e.g. `billing` is the default format in Corporate Account address book.

---
description: Configure the default checkouts process and build custom ones.
edition: commerce
---

# Configure checkout

When you work with your Commerce implementation, you can review and modify 
the checkout configuration.

!!! note "Permissions" 

>     When you modify the workflow configuration, make sure you properly set user [permissions](../permission_use_cases.md#commerce) to the checkout component.

Reviewer 2: ```suggestion
    When you modify the workflow configuration, make sure you properly set user [permissions](../../permission_use_cases.md#commerce) to the checkout component.
```

>     When you modify the workflow configuration, make sure you properly set user [permissions](../../permissions/permission_use_cases.md#commerce) to the checkout component.

Author: ```suggestion
    When you modify the workflow configuration, make sure you properly set user [permissions](permission_use_cases.md#commerce) to the checkout component.
```

## Checkout workflow

> Checkout workflow is configured like other [workflows](../../content_management/workflow/workflow.md), and each 

Author: ```suggestion
Checkout workflow is configured like other [workflows](workflow.md), and each 
```

transition represents a separate checkout step. 

By default, the checkout process is configured to render each step based on a separate 
set of libraries and templates.
Each checkout step is handled by a controller that you configure in workflow metadata.

Custom workflow implementations are defined under the `framework.workflows` key, 
and they must support the `Ibexa\Contracts\Checkout\Value\CheckoutInterface`. 
The default fallback workflow is `ibexa_checkout`, which is prepended at bundle level.
The `checkout.workflow` parameter is repository-aware.

To customize your configuration, place it in a YAML file, under the `framework.workflows.<your_workflow_name>` key, and reference it with `ibexa.repositories.<your_repository>.checkout.workflow: your_workflow_name`. 
The system can then identify which of your configured workflows handles the checkout process.

!!! note 

>     When you modify or create a controller, to ensure that no user-data is lost, extend the `Ibexa\Bundle\Checkout\Controller\AbstractStep` controller and call the `advance()` method.

Author: ```suggestion
    When you modify or create a controller, to ensure that no user data is lost, extend the `Ibexa\Bundle\Checkout\Controller\AbstractStep` controller and call the `advance()` method.
```

``` yaml 
ibexa:
    repositories:
        <repository_name>:
            checkout:
                workflow: <your_workflow_identifier> #optional, "ibexa_checkout" is used by default
                shipping_methods:
                    courier:
                        name: "Courier"
                        translation_domain: "checkout" #optional, all shipping/payment methods use this one by default
                    parcel_machine:
                        name: "Parcel machine"
                    self_pickup:
                        name: "Self pickup"
                    express_delivery:
                        name: "Express delivery"
                payment_methods:
                    credit_card:
                        name: "Credit card"
                    paypal:
                        name: "Paypal"
                    money_transfer:
                        name: "Money transfer"
```

## Shipping and billing address field format configuration 

In your implementation, you may need to create custom format configurations 
for the shipping or billing address fields, for example, to use different address 
formats based on the buyer's geographical location.

> Field formats for the billing and shipping addresses comply with the [FieldType Address](../../content_management/field_types/field_type_reference/addressfield.md#formats) specification and can be controlled with the `billing_address_format` and `shipping_address_format` flags, respectively.

Author: ```suggestion
Field formats for the billing and shipping addresses comply with the [FieldType Address](addressfield.md#formats) specification and can be controlled with the `billing_address_format` and `shipping_address_format` flags, respectively.
```

> They fallback to `billing` and `shipping` predefined formats by default:

Author: ```suggestion
They fall back to `billing` and `shipping` predefined formats by default:
```

> - `billing` is sourced from the `corporate-accounts` bundle 

Author: ```suggestion
- `billing` is part of the `ibexa/corporate-accounts` repository
```

> - `shipping` is part of the `checkout` bundle's default configuration 

Author: ```suggestion
- `shipping` is part of the `ibexa/checkout` bundle's default configuration 
```

To modify address formats you can either override the existing formats or create 
custom ones.

### Override existing Address Field Type formats 

To override the existing `billing` and `shipping` address Field Type formats, edit 

> the `../vendor/ibexa/checkout/src/bundle/Resources/config/fieldtype_address.yaml` 

Author: This is the same case as in https://github.com/ezsystems/developer-documentation/pull/1868#discussion_r1091550086, you shouldn't edit anything in vendor

file, and modify the address format, for example, by adding a tax identifier 

> to the `billing` address:

Reviewer: Again, generally speaking **the customer shouldn't edit any of our source code to override configuration**. If someone wants to have their own format, they should provide it according to Address Fieldtype doc and reference it's key using `billing_address_format` / `shipping_address_format`.

``` yaml 
formats:
    shipping:
        country:
            default:
                - region
                - locality
                - street
                - postal_code
                - email
                - phone_number
                
    billing:
        country:
            default:
                - region
                - locality
                - street
                - postal_code
                - email
                - phone_number
                - tax_number
```

!!! note

    If you are overriding the default formats and not changing their default names, 

>     there is no need to modify the keys under `ibexa.repositories.<repository_name>.checkout`.

Reviewer: Concerning the above, this note is obsolete.

### Define custom Address Field Type formats 

> To create custom Address Field Type formats to be used in Checkout, make the following changes in the configuration files. 

Reviewer: All configuration described below should be done in the project YAML files under: `ibexa_field_type_address.formats.your_format` key. In this part, I would only redirect the customer to the already existing documentation, ref: https://doc.ibexa.co/en/latest/content_management/field_types/field_type_reference/addressfield/#formats.

> To create custom Address Field Type formats to be used in Checkout, make the following changes in the configuration files. 

Author: ```suggestion
To create custom Address Field Type formats to be used in checkout, make the following changes in the configuration files. 
```

First, under under `ibexa.repositories.<repository_name>.checkout`, define custom 
format configuration keys for `billing_address_format` and `shipping_address_format`:

``` yaml 
ibexa:
    repositories:
        <repository_name>:
            checkout:
                billing_address_format: <custom_billing_fieldtype_address_format_identifier> #coming from Corporate Account "billing" by default
                shipping_address_format: <custom_shipping_fieldtype_address_format_identifier> #coming from Corporate Account "shipping" by default 

>                 customer_content_type: <your_ct_identifier_for_customer> #customer by default, it is used in registration and uses given shipping/billing addresses to pre-populate address forms in "Select Address" checkout step

Author: Best move the comments to lines above, otherwise the code block won't be readable

```

> Then, edit the `../vendor/ibexa/checkout/src/bundle/Resources/config/fieldtype_address.yaml` 

Author: As above - no editing in vendor

file, and define custom address formats, which, for example, do not include the `locality` field:

``` yaml 
formats:
    <custom_shipping_fieldtype_address_format_identifier>:
        country:
            default:
                - region
                - street
                - postal_code
                - email
                - phone_number
                
    <custom_billing_fieldtype_address_format_identifier>:
        country:
            default:
                - region
                - street
                - postal_code
                - email
                - phone_number
```

!!! note 

>     The `IbexaCheckoutExtension.php` library prepends configuration keys from the 

Author: ```suggestion
    The `IbexaCheckoutExtension` library prepends configuration keys from the 
```

>     `fieldtype_address.yaml` file with `ibexa_field_type_address`.

Reviewer: This doesn't need to be mentioned.

>     In your project, you may need to precede the `formats:` key with the 

Author: ```suggestion
    In your project, you may need to precede the `formats` key with the 
```

>     `ibexa_field_type_address:` parent key.

Reviewer: `You may` -> `you need to`.

>     `ibexa_field_type_address:` parent key.

Author: ```suggestion
    `ibexa_field_type_address` parent key.
```
